File: src/main/scala/com/vc/hard/L711.scala
package com.vc.hard

object L711 {
  def numDistinctIslands2(arr: Array[Array[Int]]): Int = {
    import scala.collection.mutable
    val island = new mutable.HashSet[String]()
    val dirs = Array((0, 1), (1, 0), (0, -1), (-1, 0))
    val coordinates = new mutable.ListBuffer[(Int, Int)]()

    val n = arr.length
    if(n == 0) return 0
    val m = arr(0).length

    def solve(x: Int, y: Int): Unit = {
      if(arr(x)(y) != -1) {
        coordinates += ((x, y))
        arr(x)(y) = -1
        var i = 0
        while(i < dirs.length) {
          val dir = dirs(i)
          val xNew = x + dir._1
          val yNew = y + dir._2
          if(xNew >= 0 && xNew < n && yNew >= 0 && yNew < m && arr(xNew)(yNew) == 1) {
            solve(xNew, yNew)
          }
          i += 1
        }
      }
    }


    object TupleOrdering extends Ordering[(Int, Int)] {
      def compare(x: (Int, Int), y: (Int, Int)): Int = {
        val cmp = x._1.compareTo(y._1)
        if(cmp == 0) x._2.compareTo(y._2) else cmp
      }
    }
    val transform = Array(
      (1, 1),
      (-1, -1),
      (1, -1),
      (-1, 1)
    )
    def translate(list: mutable.ListBuffer[(Int, Int)]) : String = {
      val transformed = new mutable.ArrayBuffer[String]()

      transform.indices.foreach(i => {
        val xTransform = transform(i)._1
        val yTransform = transform(i)._2
        val listTransformed = new mutable.ListBuffer[(Int, Int)]()
        list.foreach(xy => {
          val xNew = xy._1 * xTransform
          val yNew = xy._2 * yTransform
          listTransformed += ((xNew, yNew))
        })
        val arr = listTransformed.toArray
        scala.util.Sorting.quickSort(arr)(TupleOrdering)
        transformed += normalize(arr)
      })

      transform.indices.foreach(i => {
        val xTransform = transform(i)._1
        val yTransform = transform(i)._2
        val listTransformed = new mutable.ListBuffer[(Int, Int)]()
        list.foreach(xy => {
          val xNew = xy._1 * xTransform
          val yNew = xy._2 * yTransform
          listTransformed += ((yNew, xNew))
        })
        val arr = listTransformed.toArray
        scala.util.Sorting.quickSort(arr)(TupleOrdering)
        transformed += normalize(arr)
      })

      val resArr = transformed.toArray
      scala.util.Sorting.quickSort(resArr)
      resArr(0)
    }

    def normalize(list: Array[(Int, Int)]): String = {
      val res = new mutable.StringBuilder
      (1 until list.length).foreach(i => {
        val xy = list(i)
        res.append(s"${xy._1 - list.head._1}:${xy._2 - list.head._2}:")
      })
      res.toString
    }

    /**
            [
                [1,1,0,1,1],
                [1,0,0,0,0],
                [0,0,0,0,1],
                [1,1,0,1,1]
            ]
      */
    arr.indices.foreach(i => {
      arr(i).indices.foreach(j => {
        if(arr(i)(j) == 1) {
          coordinates.clear()
          solve(i, j)
          val key = translate(coordinates)
          island += key
        }
      })
    })
    island.size
  }
}

File: src/main/java/com/vc/hard/L711.java
package com.vc.hard;

import java.util.*;

class L711 {
    static class Point {
        int x;
        int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        @Override
        public String toString() {
            return "("+x+", "+y+")";
        }
    }

    static class PointComparator implements Comparator<Point> {
        public int compare(Point p1, Point p2) {
            int cmp = Integer.valueOf(p1.x).compareTo(p2.x);
            if(cmp == 0) return Integer.valueOf(p1.y).compareTo(p2.y);
            return cmp;
        }
    }

    public int numDistinctIslands2(int[][] grid) {
        int n = grid.length;
        if(n == 0) return 0;
        int m = grid[0].length;
        List<Point> points = new ArrayList<>();
        HashSet<String> islands = new HashSet<>();
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < m; j++) {
                if(grid[i][j] == 1) {
                    points.clear();
                    dfs(i, j, points, grid);
                    //System.out.println(points);
                    String str = transform(points);
                    //System.out.println(str);
                    islands.add(str);
                }
            }
        }
        return islands.size();
    }

    private int[][] transforms = {{1, 1}, {-1, -1}, {1, -1}, {-1, 1}};
    private String transform(List<Point> points) {
        TreeSet<String> res = new TreeSet<>();

        for(int[] transform: transforms) {
            List<Point> transformedX = new ArrayList<>();
            List<Point> transformedY = new ArrayList<>();
            int xTransform = transform[0];
            int yTransform = transform[1];
            for(Point point: points) {
                int x = point.x * xTransform;
                int y = point.y * yTransform;
                transformedX.add(new Point(x, y));
                transformedY.add(new Point(y, x));
            }
            Collections.sort(transformedX, new PointComparator());
            res.add(normalize(transformedX));

            Collections.sort(transformedY, new PointComparator());
            res.add(normalize(transformedY));
            //System.out.println("transformed: "+transformed+" str: "+str);
        }

        //System.out.println("points: "+points+" hashset: "+res);
        return res.iterator().next();
    }

    private String normalize(List<Point> points) {
        StringBuilder str = new StringBuilder();
        for(int i = 1; i < points.size(); i++) {
            int x = points.get(i).x - points.get(0).x;
            int y = points.get(i).y - points.get(0).y;
            str.append(x);
            str.append(":");
            str.append(y);
            str.append(":");
        }
        //System.out.println("points: "+points+" normalized: "+str.toString());
        return str.toString();
    }

    private int[][] dirs = {{0, 1},{1, 0},{0, -1},{-1, 0}};
    private void dfs(int x, int y, List<Point> points, int[][] grid) {
        if(grid[x][y] == 1) {
            grid[x][y] = -1;
            points.add(new Point(x, y));
            for(int[] dir: dirs) {
                int xNew = x + dir[0];
                int yNew = y + dir[1];
                if(xNew >= 0 && xNew < grid.length && yNew >= 0 && yNew < grid[0].length) {
                    dfs(xNew, yNew, points, grid);
                }
            }
        }
    }
}
